File: apps/worker/app/models/blocks.py
"""
Blocks model - Cleaned and optimized schema
"""
from datetime import datetime
from typing import Optional, Dict, Any
import enum

from sqlalchemy import String, Text, DateTime, Integer, func, ForeignKey, JSON, Enum
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base

# Define ENUM types to match Payload CMS
class BlockMediaTypeEnum(enum.Enum):
    image = "image"
    video = "video"
    gif = "gif"
    unknown = "unknown"

class BlockStatusEnum(enum.Enum):
    pending = "pending"
    fetched = "fetched"
    scraped = "scraped"
    uploaded = "uploaded"
    error = "error"


class Block(Base):
    """Blocks table - scraped content data (cleaned schema)"""
    __tablename__ = "blocks"
    
    # Primary key - using integer to match Payload
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    
    # External Reference
    external_id: Mapped[str] = mapped_column(
        String(255), 
        nullable=False, 
        unique=True,
        index=True,
        doc="Unique identifier from Savee.it"
    )
    
    # Relationships (get source_type/username via relationships)
    source_id: Mapped[int] = mapped_column(
        ForeignKey('sources.id'), 
        nullable=False, 
        index=True,
        doc="Source that discovered this block"
    )
    run_id: Mapped[int] = mapped_column(
        ForeignKey('runs.id'), 
        nullable=False, 
        index=True,
        doc="Run that scraped this block"
    )
    
    # Note: User-block relationships now handled via user_blocks junction table
    
    # Content Info
    url: Mapped[str] = mapped_column(
        Text, 
        nullable=False,
        doc="Content URL on Savee.it"
    )
    title: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Title of the content"
    )
    description: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Description of the content"
    )
    
    # Media information
    media_type: Mapped[BlockMediaTypeEnum] = mapped_column(
        Enum(BlockMediaTypeEnum, name="enum_blocks_media_type", create_type=False), 
        nullable=True,
        doc="Media type: image, video, gif, unknown"
    )
    image_url: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Original image URL"
    )
    video_url: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Original video URL"
    )
    thumbnail_url: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Thumbnail URL"
    )

    
    # Status and Processing
    status: Mapped[BlockStatusEnum] = mapped_column(
        Enum(BlockStatusEnum, name="enum_blocks_status", create_type=False), 
        default=BlockStatusEnum.pending,
        nullable=False,
        doc="Processing status: pending, fetched, scraped, uploaded, error"
    )
    
    # Comprehensive OpenGraph Metadata
    og_title: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="OpenGraph title from meta tags"
    )
    og_description: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="OpenGraph description from meta tags"
    )
    og_image_url: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="OpenGraph image URL from meta tags"
    )
    og_url: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="OpenGraph canonical URL"
    )
    source_api_url: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Savee API endpoint for source resolution"
    )
    saved_at: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="ISO timestamp when item was scraped"
    )

    # Rich Metadata for Filtering/Search
    color_hexes: Mapped[Dict[str, Any]] = mapped_column(
        JSON, 
        nullable=True,
        doc="Array of hex color codes extracted from image"
    )
    ai_tags: Mapped[Dict[str, Any]] = mapped_column(
        JSON, 
        nullable=True,
        doc="AI-generated descriptive tags for content"
    )
    colors: Mapped[Dict[str, Any]] = mapped_column(
        JSON, 
        nullable=True,
        doc="Array of RGB color values"
    )
    links: Mapped[Dict[str, Any]] = mapped_column(
        JSON, 
        nullable=True,
        doc="Links extracted from item sidebar (includes original source URLs)"
    )
    metadata_: Mapped[Dict[str, Any]] = mapped_column(
        "metadata",  # Specify the actual column name
        JSON, 
        nullable=True,
        doc="Complete sidebar info and other metadata"
    )
    
    # Persisted fields used for CMS filtering/search (optional)
    origin_text: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
        doc="Computed origin for filters (home | pop | username)"
    )
    saved_by_usernames: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
        doc="Comma-separated usernames who saved this block"
    )
    
    # Storage
    r2_key: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="R2 storage key for uploaded media"
    )
    
    # Error handling
    error_message: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Error message if processing failed"
    )
    
    # Timestamps (standard Payload)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        doc="When this block was created"
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
        doc="When this block was last updated"
    )

    # Relationships
    source = relationship("Source")
    run = relationship("Run")

    def __repr__(self) -> str:
        return f"<Block(id={self.id}, external_id='{self.external_id}', media_type='{self.media_type}', status='{self.status}')>"

File: apps/cms/src/collections/Blocks.ts
import type { CollectionConfig } from "payload";

let didBackfillFilters = false;

export const Blocks: CollectionConfig = {
  slug: "blocks",
  hooks: {
    afterRead: [
      async ({ req }) => {
        if (didBackfillFilters) return;
        didBackfillFilters = true;
        try {
          const db = (req.payload.db as any).pool;
          await db.query(
            `UPDATE blocks b
           SET origin_text = COALESCE(origin_text,
             CASE WHEN s.source_type = 'user' THEN s.username ELSE s.source_type END),
             saved_by_usernames = COALESCE(saved_by_usernames, sub.usernames)
           FROM sources s
           LEFT JOIN (
             SELECT ub.block_id, string_agg(u.username, ',') AS usernames
             FROM user_blocks ub
             JOIN savee_users u ON u.id = ub.user_id
             GROUP BY ub.block_id
           ) AS sub ON sub.block_id = b.id
           WHERE b.source_id = s.id`
          );
        } catch {}
      },
    ],
  },
  admin: {
    useAsTitle: "title",
    defaultColumns: [
      "preview",
      "title",
      "media_type", // Match database column name
      "origin", // Computed: home | pop | username
      "saved_by",
      "status",
      "source",
      "run",
      "createdAt",
    ],
    description: "Individual scraped content blocks from Savee.it",
    listSearchableFields: ["title", "url", "og_title", "og_description"],
    listFilterableFields: [
      "origin_text",
      "saved_by_usernames",
      "media_type",
      "status",
      "source",
      "run",
    ],
  },
  access: {
    read: () => true,
    create: () => true,
    update: () => true,
    delete: () => true,
  },
  fields: [
    // Preview (admin-only UI field)
    {
      name: "preview",
      type: "ui",
      label: "Preview",
      admin: {
        components: {
          Cell: "@/components/BlockPreviewCell",
        },
      },
    },
    // Saved by usernames
    {
      name: "saved_by",
      type: "ui",
      label: "Saved By",
      admin: {
        components: {
          Cell: "@/components/BlockUsersCell",
        },
      },
    },
    // Appears In (home, pop, users consolidated)
    {
      name: "origin",
      type: "ui",
      label: "Appears In",
      admin: {
        description: "Where this block appears: Home, Popular, and Users",
        components: {
          Cell: "@/components/MultiOriginCell",
        },
      },
    },
    // Persisted origin (filterable/searchable)
    {
      name: "origin_text",
      type: "text",
      label: "Savee User / Origin (text)",
      admin: {
        description: "Persisted origin for filters (home | pop | username)",
      },
    },
    // External Reference
    {
      name: "external_id", // Match database column name
      type: "text",
      required: true,
      unique: true,
      label: "External ID",
      admin: {
        description: "Unique identifier from Savee.it",
      },
    },
    // Persisted Saved By usernames (comma-separated), filterable/searchable
    {
      name: "saved_by_usernames",
      type: "text",
      label: "Saved By (usernames)",
      admin: {
        description: "Comma-separated usernames who saved this block",
      },
    },

    // Relationships (source info available via relationships)
    {
      name: "source",
      type: "relationship",
      relationTo: "sources",
      required: true,
      label: "Source",
    },
    {
      name: "run",
      type: "relationship",
      relationTo: "runs",
      required: true,
      label: "Run",
    },
    {
      name: "savee_user",
      type: "relationship",
      relationTo: "savee_users",
      label: "Savee User",
      admin: {
        description: "SaveeUser profile for user content organization",
      },
    },

    // Content Info
    {
      name: "url",
      type: "text",
      required: true,
      label: "Content URL",
    },
    {
      name: "title",
      type: "text",
      label: "Title",
    },
    {
      name: "description",
      type: "textarea",
      label: "Description",
    },

    // Media Information
    {
      name: "media_type", // Match database column name
      type: "select",
      label: "Media Type",
      options: [
        { label: "Image", value: "image" },
        { label: "Video", value: "video" },
        { label: "GIF", value: "gif" },
        { label: "Unknown", value: "unknown" },
      ],
    },
    // (No extra persisted filter-only fields; use origin_text and saved_by_usernames in the filter UI.)
    {
      name: "image_url", // Match database column name
      type: "text",
      label: "Image URL",
    },
    {
      name: "video_url", // Match database column name
      type: "text",
      label: "Video URL",
    },
    {
      name: "thumbnail_url", // Match database column name
      type: "text",
      label: "Thumbnail URL",
    },

    // Status and Processing
    {
      name: "status",
      type: "select",
      required: true,
      defaultValue: "pending",
      label: "Status",
      options: [
        { label: "Pending", value: "pending" },
        { label: "Fetched", value: "fetched" },
        { label: "Scraped", value: "scraped" },
        { label: "Uploaded", value: "uploaded" },
        { label: "Error", value: "error" },
      ],
    },

    // Comprehensive OpenGraph Metadata
    {
      name: "og_title", // Match database column name
      type: "text",
      label: "OG Title",
      admin: {
        description: "OpenGraph title from meta tags",
      },
    },
    {
      name: "og_description", // Match database column name
      type: "textarea",
      label: "OG Description",
      admin: {
        description: "OpenGraph description from meta tags",
      },
    },
    {
      name: "og_image_url", // Match database column name
      type: "text",
      label: "OG Image URL",
      admin: {
        description: "OpenGraph image URL from meta tags",
      },
    },
    {
      name: "og_url", // Match database column name
      type: "text",
      label: "OG URL",
      admin: {
        description: "OpenGraph canonical URL",
      },
    },
    {
      name: "source_api_url", // Match database column name
      type: "text",
      label: "Source API URL",
      admin: {
        description: "Savee API endpoint for source resolution",
      },
    },
    {
      name: "saved_at", // Match database column name
      type: "text",
      label: "Saved At",
      admin: {
        description: "ISO timestamp when item was scraped",
      },
    },

    // Rich Metadata for Filtering/Search
    {
      name: "color_hexes", // Match database column name
      type: "json",
      label: "Color Hex Codes",
      admin: {
        description: "Array of hex color codes extracted from image",
      },
    },
    {
      name: "ai_tags", // Match database column name
      type: "json",
      label: "AI Generated Tags",
      admin: {
        description: "AI-generated descriptive tags for content",
      },
    },
    {
      name: "colors",
      type: "json",
      label: "RGB Colors",
      admin: {
        description: "Array of RGB color values",
      },
    },
    {
      name: "links",
      type: "json",
      label: "Sidebar Links",
      admin: {
        description: "Links extracted from item sidebar",
      },
    },
    {
      name: "metadata",
      type: "json",
      label: "Additional Metadata",
      admin: {
        description: "Complete sidebar info and other metadata",
      },
    },

    // Storage
    {
      name: "r2_key", // Match database column name
      type: "text",
      label: "R2 Storage Key",
      admin: {
        readOnly: true,
      },
    },

    // Error Handling
    {
      name: "errorMessage",
      type: "textarea",
      label: "Error Message",
      admin: {
        condition: (data) => data.status === "error",
      },
    },
  ],
};
